ocr/tess/crop_morphology.py

```diff
@@ -230,69 +302,164 @@ def downscale_image(im, max_dim=2048):
     maxed_rows = rank_filter(edges, -4, size=(1, 20))
     maxed_cols = rank_filter(edges, -4, size=(20, 1))
     debordered = np.minimum(np.minimum(edges, maxed_rows), maxed_cols)
     edges = debordered
+    # Image.fromarray(edges).show()
 
     contours = find_components(edges)
     if len(contours) == 0:
-        print '%s -> (no text!)' % path
+        sys.stderr.write('%s: no text!\n' % path)
         return
 
-    crop = find_optimal_components_subset(contours, edges)
-    crop = pad_crop(crop, contours, edges, border_contour)
-
-    crop = [int(x / scale) for x in crop]  # upscale to the original image size.
-    #draw = ImageDraw.Draw(im)
-    #c_info = props_for_contours(contours, edges)
-    #for c in c_info:
-    #    this_crop = c['x1'], c['y1'], c['x2'], c['y2']
-    #    draw.rectangle(this_crop, outline='blue')
-    #draw.rectangle(crop, outline='red')
-    #im.save(out_path)
-    #draw.text((50, 50), path, fill='red')
-    #orig_im.save(out_path)
-    #im.show()
-    text_im = orig_im.crop(crop)
-    text_im.save(out_path)
-    print '%s -> %s' % (path, out_path)
+    if not border_only:
+        crop = find_optimal_components_subset(contours, edges, beta)
+        crop = pad_crop(crop, contours, edges, border_contour)
+    else:
+        if border_contour is not None:
+            c_info = props_for_contours([border_contour], edges)
+            c = c_info[0]
+            crop = c['x1'], c['y1'], c['x2'], c['y2']
+        else:
+            crop = 0, 0, im.width, im.height
+
+    if stroke:
+        im = im.convert('RGB')
+        draw = ImageDraw.Draw(im)
+        c_info = props_for_contours(contours, edges)
+        for c in c_info:
+            this_crop = c['x1'], c['y1'], c['x2'], c['y2']
+            draw.rectangle(this_crop, outline='blue')
+        draw.rectangle(crop, outline='red')
+        if border_contour is not None:
+            c_info = props_for_contours(border_contours, edges)
+            for i, c in enumerate(c_info):
+                this_crop = c['x1'], c['y1'], c['x2'], c['y2']
+                draw.rectangle(this_crop, outline='green', width=4 if i == 0 else 2)
+
+        if stamp_contour is not None:
+            draw.line(
+                [(pt[0][0], pt[0][1]) for pt in stamp_contour], fill="hotpink", width=4
+            )
+        out_im = im
+    else:
+        crop = [int(x / scale) for x in crop]  # upscale to the original image size.
+        out_im = orig_im.crop(crop)
+
+    # draw.text((50, 50), path, fill='red')
+    # orig_im.save(out_path)
+    # im.show()
+    out_im.save(out_path)
+    print('%s -> %s' % (path, out_path))
 
 
 if __name__ == '__main__':
-    if len(sys.argv) == 2 and '*' in sys.argv[1]:
-        files = glob.glob(sys.argv[1])
-        random.shuffle(files)
-    else:
-        files = sys.argv[1:]
+    parser = argparse.ArgumentParser(description='Crop images to just the text')
+    parser.add_argument(
+        "--beta",
+        default=1.0,
```

Comment:
Using beta=2 results in more conservative cropping, which is probably the right tradeoff. You can't OCR any text that you crop out!